### PR TITLE
Fix the context creation without config

### LINKED ISSUE #582: Context is initialized when IdealState is existed but config is missing
In WorkflowDispatcher.java there is a function called getOrInitializeWorkflowContext that creates the context without checking the config. In this case, if workflow is expired and IdealState is still existed in cache or ZooKeeper, getOrInitializeWorkflowContext creates a new context which is not desirable.

### PR DESCRIPTION
**Pull request by alirezazamani**
### Issues
- [x] My PR addresses the following Helix issues and references them in the PR title:
Fixes #582.

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

An if statement is added in the getOrInitializeWorkflowContext fucntion the checks whether config is null or not. If config is null, this function will not initialize the context.

### Tests

- [x] The following is the result of the "mvn test" command on the appropriate module:
A new test has been added. (TestWorkflowContextWithoutConfig)

Test Result 1: mvn test

[ERROR] Tests run: 883, Failures: 2, Errors: 0, Skipped: 1, Time elapsed: 3,435.407 s <<< FAILURE! - in TestSuite
[ERROR] testTagSetIncorrect(org.apache.helix.integration.TestAlertingRebalancerFailure)  Time elapsed: 180.079 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.integration.TestAlertingRebalancerFailure.checkResourceBestPossibleCalFailureState(TestAlertingRebalancerFailure.java:310)
	at org.apache.helix.integration.TestAlertingRebalancerFailure.testTagSetIncorrect(TestAlertingRebalancerFailure.java:174)

[ERROR] testTaskPerformanceMetrics(org.apache.helix.monitoring.mbeans.TestTaskPerformanceMetrics)  Time elapsed: 2.219 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.monitoring.mbeans.TestTaskPerformanceMetrics.testTaskPerformanceMetrics(TestTaskPerformanceMetrics.java:118)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestAlertingRebalancerFailure.testTagSetIncorrect:174->checkResourceBestPossibleCalFailureState:310 expected:<true> but was:<false>
[ERROR]   TestTaskPerformanceMetrics.testTaskPerformanceMetrics:118 expected:<true> but was:<false>
[INFO] 
[ERROR] Tests run: 883, Failures: 2, Errors: 0, Skipped: 1
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  57:20 min
[INFO] Finished at: 2019-11-01T10:34:40-07:00
[INFO] ------------------------------------------------------------------------

Test Result 2: mvn test -Dtest="TestAlertingRebalancerFailure,TestTaskPerformanceMetrics"
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 19.918 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 4, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  24.422 s
[INFO] Finished at: 2019-11-01T11:05:16-07:00
[INFO] ------------------------------------------------------------------------

### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/integration/task/TestWorkflowContextWithoutConfig.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    // workflow
+    Assert.assertNotNull(_driver.getWorkflowConfig(workflowName1));
+    Assert.assertNotNull(_driver.getWorkflowContext(workflowName1));
+    Assert.assertNotNull(_admin.getResourceIdealState(CLUSTER_NAME, workflowName1));
+
+    // Wait until workflow is completed.
+    _driver.pollForWorkflowState(workflowName1, TaskState.COMPLETED);
+
+    String idealStatePath = "/" + CLUSTER_NAME + "/IDEALSTATES/" + workflowName1;
+    ZNRecord record = _manager.getHelixDataAccessor().getBaseDataAccessor().get(idealStatePath,
```
Comment (narendly) [inline helix-core/src/test/java/org/apache/helix/integration/task/TestWorkflowContextWithoutConfig.java]: Could this possibly be null?

Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestWorkflowContextWithoutConfig.java]: Idealstate will be deleted (null) when workflow expired. If the workflow is in progress, IdealState should not be null. Just to make sure, I moved this part before polling for COMPLETED state. Also I added an assert to make sure record is not null. Thanks for the comment.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestWorkflowContextWithoutConfig.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    // expired.
+    boolean workflowExpired = TestHelper.verify(() -> {
+      WorkflowContext wCtx = _driver.getWorkflowContext(workflowName1);
+      WorkflowConfig wCfg = _driver.getWorkflowConfig(workflowName1);
+      IdealState idealState = _admin.getResourceIdealState(CLUSTER_NAME, workflowName1);
+      return (wCtx == null && wCfg == null && idealState == null);
+    }, 30 * 1000);
+    Assert.assertTrue(workflowExpired);
+
+    // Write ideaState to ZooKeeper
```
[the remark this answers is not in the document]
Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestWorkflowContextWithoutConfig.java]: Good catch. Fixed.

File: helix-core/src/test/java/org/apache/helix/integration/task/TestWorkflowContextWithoutConfig.java (hunk 1 above)
Comment (narendly) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestWorkflowContextWithoutConfig.java]: Doesn't the idealstate for the workflow get dropped after it completed? Or am I thinking of jobs' IdealStates?

Comment (alirezazamani) [reply helix-core/src/test/java/org/apache/helix/integration/task/TestWorkflowContextWithoutConfig.java]: The idealstate of the job will be dropped when job is failed or completed. For the workflow idealstate is deleted when workflow is expired.

**Comment by alirezazamani** [comment]
> Looks good, but in your tests I see a lot of duplicate code. We'll perhaps let this one slide, but if you ever need to write a similar test, consider refactoring so that we aren't copying and pasting the boilerplate code over and over.

Thanks Hunter. Sure.

**Comment by alirezazamani** [comment]
This PR is ready to be merged, approved by @narendly.

Final commit message:
Title: 
Fix the context creation without config

Body:
Create context only if config is existed.
An if statement is added in the getOrInitializeWorkflowContext function:
It checks whether config is null or not.
If config is null, this function will not initialize the context.

A test has been added to check this functionality.

**Pull request merged**